### PR TITLE
Add portable select() reactor backend for POSIX platforms

### LINKED ISSUE #61: Implement select reactor
## Summary

The corosio library currently only supports epoll (Linux) and IOCP (Windows) backends. Adding a `select` reactor would provide a portable fallback for platforms where epoll is unavailable, including older Linux kernels (pre-2.6), QNX Neutrino, AIX, HP-UX, Tru64, and potentially as a debugging/testing backend.

### Platform Coverage

The `select()` system call is part of the POSIX standard and is available on virtually all Unix-like systems:

| Platform | Native Backend | select Available |
|----------|----------------|------------------|
| Linux 2.6+ | epoll | Yes |
| Linux 2.4 | - | Yes |
| macOS/FreeBSD | kqueue (future) | Yes |
| Solaris | /dev/poll (future) | Yes |
| QNX Neutrino | - | Yes |
| AIX, HP-UX, Tru64 | - | Yes |

### Use Cases

1. **Portability**: Enable corosio on platforms without epoll
2. **Debugging**: Simpler implementation for isolating issues
3. **Testing**: Validate backend-agnostic code paths
4. **Embedded systems**: Some embedded Linux builds lack epoll support

## Current State

**Existing backends**: `src/corosio/src/detail/epoll/` and `src/corosio/src/detail/iocp/`

The epoll backend consists of:
- `scheduler.hpp/cpp` - Event loop with epoll_wait, work queuing, thread coordination
- `sockets.hpp/cpp` - Socket operations and async I/O
- `op.hpp` - Operation types for socket events
- `resolver_service.hpp` - DNS resolution (stub)

## Boost.Asio Reference Implementation

Boost.Asio's `select_reactor` provides a well-tested reference. Key implementation details:

**Location**: `boost/asio/detail/select_reactor.hpp` and `select_reactor.ipp`

### Architecture

```
┌─────────────────────────────────────────────────────────────┐
│                     select_reactor                          │
├─────────────────────────────────────────────────────────────┤
│  fd_set_adapter fd_sets_[max_select_ops]  (read/write/err) │
│  reactor_op_queue<socket_type> op_queue_[max_ops]          │
│  select_interrupter interrupter_                           │
│  mutex mutex_                                               │
└─────────────────────────────────────────────────────────────┘
```

### Key Components

1. **fd_set_adapter**: Wrapper around `fd_set` that handles the `FD_SETSIZE` limit and provides a cleaner interface for adding/removing descriptors.

2. **reactor_op_queue**: Per-descriptor operation queue. Operations are registered for read, write, or error events and executed when select returns.

3. **select_interrupter**: Mechanism to wake the select call (typically a pipe or self-connected socket). Writing to the interrupter causes select to return immediately.

4. **Operation registration**:
   ```cpp
   void start_op(int op_type, socket_type descriptor,
                 reactor_op* op, bool is_continuation);
   ```

5. **Event loop**:
   ```cpp
   void run(long usec, op_queue<operation>& ops)
   {
       // Build fd_sets from registered operations
       // Call select() with timeout
       // For each ready descriptor, dequeue and execute operations
   }
   ```

### FD_SETSIZE Limitation

The `select()` call has a compile-time limit (`FD_SETSIZE`, typically 1024) on the number of file descriptors it can monitor. Boost.Asio handles this by:

1. Tracking the maximum fd in use
2. Returning an error if registration exceeds `FD_SETSIZE`
3. Documenting the limitation

For corosio, we accept the limitation since it is sufficient for many use cases, and platforms where it matters have better backends available (epoll, kqueue).

### Why select() Instead of poll()

A natural question is why not use `poll()` to avoid the `FD_SETSIZE` limitation. Boost.Asio chose `select()` as the portable fallback for several reasons:

1. **Windows compatibility**: Windows didn't have `poll()` until Vista (2007) when `WSAPoll` was added. Before that, `select()` was the only portable option.

2. **WSAPoll was buggy**: Microsoft had a known bug ("Windows 8 Bugs 309411") where `WSAPoll` didn't report failed TCP connections properly. It was marked "Won't Fix" and wasn't fully resolved until Windows 10 version 2004 (2020).

3. **Maximum portability**: `select()` works on all Windows versions (including CE) and virtually all POSIX systems.

4. **No middle-ground benefit**: On platforms where `FD_SETSIZE` is a real limitation, superior alternatives already exist:
   - Linux → epoll
   - BSD/macOS → kqueue
   - Solaris → /dev/poll

The trade-off is:
- `select()` = maximum portability, but FD_SETSIZE limit
- `poll()` = no FD_SETSIZE limit, but Windows-problematic
- Platform-specific (epoll/kqueue) = best performance, no fd limits

Since you either need maximum portability (select) or you're on a platform with something better anyway, the middle ground of `poll()` provides little value.

### Directory Structure

```
src/corosio/src/detail/select/
├── scheduler.hpp
├── scheduler.cpp
├── sockets.hpp
├── sockets.cpp
├── op.hpp
└── fd_set_adapter.hpp
```

**Note**: No `resolver_service.hpp` or signal implementation needed — the select backend reuses the existing POSIX implementations (see "Reusing POSIX Services" section below).

### Phase 1: Core Scheduler

Create `select_scheduler` with:

1. **fd_set management**:
   ```cpp
   class fd_set_adapter
   {
   public:
       void set(int fd);
       void clear(int fd);
       bool is_set(int fd) const;
       fd_set* get() { return &set_; }
       int max_fd() const { return max_fd_; }
   private:
       fd_set set_;
       int max_fd_ = -1;
   };
   ```

2. **Operation queues** (per-fd, per-event-type):
   ```cpp
   std::unordered_map<int, scheduler_op*> read_ops_;
   std::unordered_map<int, scheduler_op*> write_ops_;
   std::unordered_map<int, scheduler_op*> error_ops_;
   ```

3. **Interrupter** (pipe-based):
   ```cpp
   class select_interrupter
   {
   public:
       select_interrupter();
       ~select_interrupter();
       void interrupt();
       void reset();
       int read_descriptor() const { return read_fd_; }
   private:
       int read_fd_;
       int write_fd_;
   };
   ```

4. **Event loop**:
   ```cpp
   void run_reactor(std::unique_lock<std::mutex>& lock)
   {
       // Build fd_sets from registered operations
       fd_set_adapter read_fds, write_fds, except_fds;
       int max_fd = build_fd_sets(read_fds, write_fds, except_fds);

       // Add interrupter to read set
       read_fds.set(interrupter_.read_descriptor());
       max_fd = std::max(max_fd, interrupter_.read_descriptor());

       lock.unlock();

       timeval tv = calculate_timeout();
       int result = ::select(max_fd + 1,
                             read_fds.get(),
                             write_fds.get(),
                             except_fds.get(),
                             &tv);

       lock.lock();

       if (result > 0)
           process_ready_descriptors(read_fds, write_fds, except_fds);
   }
   ```

### Phase 2: Socket Operations

Adapt epoll socket operations for select:

1. **Registration**: Instead of `epoll_ctl(EPOLL_CTL_ADD)`, add fd to appropriate operation map
2. **Modification**: Update which events are monitored
3. **Completion**: Remove fd from maps when operation completes

### Phase 3: Integration

**Backend selection** (`config_backend.hpp`):

The select backend serves as the automatic fallback when platform-specific backends are unavailable. No user-configurable compile flags are needed — the backend is selected based on platform detection:

```cpp
#if defined(__linux__)
#  include "epoll/scheduler.hpp"
   namespace corosio::detail { using scheduler = epoll_scheduler; }
#elif defined(_WIN32)
#  include "iocp/scheduler.hpp"
   namespace corosio::detail { using scheduler = iocp_scheduler; }
#else
   // Fallback for all other POSIX platforms (QNX, AIX, HP-UX, etc.)
#  include "select/scheduler.hpp"
   namespace corosio::detail { using scheduler = select_scheduler; }
#endif
```

This approach:
- Keeps the build configuration simple with no user-facing options
- Automatically uses the best available backend per platform
- Provides select as the fallback for any POSIX system without a specialized backend

## Reusing POSIX Services

The select backend should reuse the existing POSIX signal and resolver implementations rather than creating backend-specific versions.

### Signals (`posix/signals.hpp/cpp`)

The POSIX signal implementation uses `sigaction()` with C signal handlers and is scheduler-agnostic via the abstract `scheduler` interface. During scheduler initialization:

```cpp
// In select_scheduler constructor
signal_svc_ = &get_signal_service(ctx, *this);
```

### Resolver (`posix/resolver_service.hpp/cpp`)

The POSIX resolver uses `getaddrinfo()` on worker threads — the standard approach for all POSIX systems since there's no async DNS API. The header explicitly states:

```cpp
// This implementation works for all POSIX backends (epoll, kqueue, io_uring, poll)
```

During scheduler initialization:

```cpp
// In select_scheduler constructor
resolver_svc_ = &get_resolver_service(ctx, *this);
```

### Scheduler Interface Requirements

The select scheduler must implement the abstract `scheduler` interface (`include/boost/corosio/detail/scheduler.hpp`):

```cpp
struct scheduler
{
    virtual void post(capy::coro) const = 0;
    virtual void post(scheduler_op*) const = 0;
    virtual void on_work_started() noexcept = 0;
    virtual void on_work_finished() noexcept = 0;
    virtual void work_started() const noexcept = 0;
    virtual void work_finished() const noexcept = 0;
    virtual bool running_in_this_thread() const noexcept = 0;
    virtual void stop() = 0;
    virtual bool stopped() const noexcept = 0;
    virtual void restart() = 0;
    virtual std::size_t run() = 0;
    virtual std::size_t run_one() = 0;
    virtual std::size_t wait_one(long usec) = 0;
    virtual std::size_t poll() = 0;
    virtual std::size_t poll_one() = 0;
};
```

### Core Infrastructure

- [ ] Create `select/fd_set_adapter.hpp` - fd_set wrapper with max_fd tracking
- [ ] Create `select/select_interrupter.hpp/cpp` - Pipe-based wakeup mechanism
- [ ] Create `select/scheduler.hpp/cpp` - Event loop implementation

### Socket Support

- [ ] Create `select/op.hpp` - Operation types (can likely reuse from epoll)
- [ ] Create `select/sockets.hpp/cpp` - Socket async operations
- [ ] Implement async_accept, async_connect, async_read, async_write

### Integration

- [ ] Update `config_backend.hpp` with select backend as fallback
- [ ] Initialize POSIX services in scheduler constructor (`get_signal_service`, `get_resolver_service`)

### Testing

- [ ] Verify all existing tests pass with select backend
- [ ] Add select-specific tests for FD_SETSIZE boundary conditions
- [ ] Test on non-Linux POSIX platform if available

## Files to Create/Modify

| File | Action | Description |
|------|--------|-------------|
| `src/corosio/src/detail/select/` | Create | New backend directory |
| `src/corosio/src/detail/select/scheduler.hpp` | Create | Scheduler class (implements `scheduler` interface) |
| `src/corosio/src/detail/select/scheduler.cpp` | Create | Event loop, POSIX service initialization |
| `src/corosio/src/detail/select/sockets.hpp` | Create | Socket operation declarations |
| `src/corosio/src/detail/select/sockets.cpp` | Create | Socket operation implementations |
| `src/corosio/src/detail/select/op.hpp` | Create | Operation types |
| `src/corosio/src/detail/select/fd_set_adapter.hpp` | Create | fd_set wrapper |
| `src/corosio/src/detail/select/select_interrupter.hpp` | Create | Wakeup mechanism |
| `src/corosio/src/detail/select/select_interrupter.cpp` | Create | Interrupter implementation |
| `src/corosio/src/detail/config_backend.hpp` | Modify | Add select as fallback backend |

**Note**: No resolver or signal files needed — reuses `posix/resolver_service` and `posix/signals`.

## Known Limitations

1. **FD_SETSIZE**: Maximum ~1024 file descriptors (platform-dependent)
2. **Performance**: O(n) per select call vs O(1) for epoll
3. **No edge-triggered mode**: select is always level-triggered

## References

- Boost.Asio select_reactor: `boost/asio/detail/select_reactor.hpp`
- Boost.Asio select_reactor implementation: `boost/asio/detail/impl/select_reactor.ipp`
- Boost.Asio fd_set_adapter: `boost/asio/detail/fd_set_adapter.hpp`
- Boost.Asio pipe_select_interrupter: `boost/asio/detail/pipe_select_interrupter.hpp`
- POSIX select specification: [pubs.opengroup.org](https://pubs.opengroup.org/onlinepubs/9699919799/functions/select.html)
- Existing epoll backend: `src/corosio/src/detail/epoll/`

### PR DESCRIPTION
**Pull request by sgerbino**
Implement a select()-based I/O multiplexing backend that provides runtime backend selection on platforms where multiple backends are available. On Linux, users can now choose between epoll_context (best performance) and select_context (portable) at runtime.

Architecture changes:
- Add abstract socket_service and acceptor_service base classes to enable runtime polymorphism between epoll and select backends
- Split epoll socket/acceptor services to avoid diamond inheritance with execution_context::service
- Update socket.cpp and acceptor.cpp to use abstract service interfaces
- Delete config_backend.hpp entirely: use platform macros directly (_WIN32, __linux__, !defined(_WIN32)) instead of custom macros

Key design decisions:
- Self-pipe pattern for reactor interruption (portable, no eventfd)
- FD_SETSIZE limit enforced with graceful error (no undefined behavior)
- Single-reactor thread model matching epoll for consistent behavior
- Level-triggered semantics (inherent to select)

Bug fixes (applied to both select and epoll backends):
- Fix race condition between register_fd() and cancel_single_op() that could leave fds in the reactor with dangling pointers. Reordered to call register_fd() before setting the registered flag.
- Add defensive deregister in close_socket() to clean up any leaked fd entries from edge-case race conditions
- Add null state check in acceptor service constructor to fail fast if socket service is not installed

Test updates:
- Add multi-backend test parameterization to acceptor, socket, timer, and signal_set tests using template pattern
- Tests run for both io_context and select_context on POSIX platforms
- Fix testWriteAfterPeerClose to write more data for macOS compatibility (macOS has larger socket buffers, needs more data to trigger EPIPE)

Benchmark updates:
- Add --backend option to io_context_bench for runtime backend selection
- Add --list option to show available backends on the platform
- Benchmarks can now compare epoll vs select performance on Linux

CI updates:
- Enable macOS builds using select backend (kqueue planned for future)
- Update runner from macos-26 to macos-15
- Add vcpkg.json and vcpkg paths setup for macOS builds
- Add -fexperimental-library flag for std::stop_token support on macOS

Build fixes:
- Link CoreFoundation and Security frameworks for WolfSSL on macOS (vcpkg-built WolfSSL uses Apple certificate APIs)

Resolves #61.

## Summary by CodeRabbit

* **New Features**
  * Added `cancel()` method to socket and acceptor classes for operation cancellation.
  * Added select-based I/O backend for non-Linux POSIX platforms (macOS, BSD).
  * Enhanced macOS build support with new compiler configurations.

* **Improvements**
  * Improved platform-specific I/O backend selection at runtime.
  * Better support for diverse POSIX environments with automatic backend selection.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: include/boost/corosio/socket.hpp [hunk 1]
```diff
@@ -101,6 +101,17 @@ class BOOST_COROSIO_DECL socket : public io_stream
...
+        /** Request cancellation of pending asynchronous operations.
+
+            When a backend supports explicit cancellation it will override
+            cancel() and all outstanding operations complete with
+            operation_canceled error. The default implementation does nothing
+            for backends that don't support explicit cancellation, so the
+            postcondition of operations completing with operation_canceled
+            is not universal unless the backend overrides cancel().
+        */
+        virtual void cancel() noexcept {}
```
Comment (vinniefalco) [inline include/boost/corosio/socket.hpp]: why isn't this pure virtual

File: src/corosio/src/detail/select/sockets.hpp [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+    int send_buffer_size(system::error_code& ec) const noexcept override;
+
+    system::error_code set_linger(bool enabled, int timeout) noexcept override;
+    socket::linger_options linger(system::error_code& ec) const noexcept override;
+
+    endpoint local_endpoint() const noexcept override { return local_endpoint_; }
+    endpoint remote_endpoint() const noexcept override { return remote_endpoint_; }
+    bool is_open() const noexcept { return fd_ >= 0; }
+    void cancel() noexcept override;
+    void cancel_single_op(select_op& op) noexcept;
```
[the remark this answers is not in the document]
Comment (sgerbino) [reply src/corosio/src/detail/select/sockets.hpp:143]: False positive?

File: src/corosio/src/detail/select/op.hpp [hunk 3]
```diff
@@ -0,0 +1,397 @@
...
+struct select_op : scheduler_op
+{
+    struct canceller
+    {
+        select_op* op;
+        void operator()() const noexcept;
+    };
+
+    capy::coro h;
+    capy::executor_ref d;
```
Comment (vinniefalco) [inline src/corosio/src/detail/select/op.hpp:117]: should be
```
capy::executor_ref ex;
```

File: src/corosio/src/detail/select/op.hpp [hunk 4]
```diff
@@ -0,0 +1,397 @@
...
+            else if (errn != 0)
+                *ec_out = make_err(errn);
+            else if (is_read_operation() && bytes_transferred == 0)
+                *ec_out = capy::error::eof;
+        }
+
+        if (bytes_out)
+            *bytes_out = bytes_transferred;
+
+        auto saved_d = d;
```
Comment (vinniefalco) [inline src/corosio/src/detail/select/op.hpp:170]: This needs a comment that we are moving things to the stack because we are about to destroy the frame. Also I prefer seeing the whole type at sensitive places like this:
```
executor_ref ex( std::move(ex_) );
```

File: src/corosio/src/detail/select/op.hpp [hunk 5]
```diff
@@ -0,0 +1,397 @@
...
+
+    void start(std::stop_token token, select_acceptor_impl* impl)
+    {
+        cancelled.store(false, std::memory_order_release);
+        stop_cb.reset();
+        socket_impl_ = nullptr;
+        acceptor_impl_ = impl;
+
+        if (token.stop_possible())
+            stop_cb.emplace(token, canceller{this});
```
Comment (vinniefalco) [inline src/corosio/src/detail/select/op.hpp:219]: why not just always store the token and always check it
optionals are an eyesore
